compiler/rustc_mir_transform/src/dataflow_const_prop.rs

```diff
@@ -194,6 +211,19 @@ impl<'tcx> ValueAnalysis<'tcx> for ConstAnalysis<'_, 'tcx> {
         state: &mut State<Self::Value>,
     ) -> ValueOrPlace<Self::Value> {
         let val = match rvalue {
+            Rvalue::Len(place) => {
+                let place_ty = place.ty(self.local_decls, self.tcx);
+                if let ty::Array(_, len) = place_ty.ty.kind() {
+                    ConstantKind::Ty(*len)
+                        .eval(self.tcx, self.param_env)
+                        .try_to_scalar_int()
+                        .map_or(FlatSet::Top, FlatSet::Elem)
+                } else if let [ProjectionElem::Deref] = place.projection[..] {
+                    state.get_len(place.local.into(), self.map())
+                } else {
+                    FlatSet::Top
+                }
```

Comment:
There's another option. If the slice is from a constant, we could even read the length from the constant, without it needing to be derived from an array.

Reply:
In that case, we'd still have `Len(*_local)` where `_local` holds the slice. For that case, I'd rather directly implement general assignments `x = const` where the RHS is not a scalar.